### PR TITLE
perf: optimize patch helper hot paths

### PR DESCRIPTION
**Pull request by sozercan**
## Summary

- Add focused microbenchmarks for report parsing, package-manager validation/parsing, platform/buildkit helpers, and bulk patch helper paths.
- Optimize hot paths in Trivy fixed-version selection, package validation/manifest parsing, platform filtering, local manifest descriptor parsing, and bulk tag/report handling.
- Preserve security-sensitive behavior around package-name validation, patch tag parsing, and report JSON compatibility; add regression tests for signed patch-tag suffixes and escaped slash `ArtifactName` values.

## Impact

Representative benchmark improvements from the final runs include:

- Trivy fixed-version selection: ~83% faster, allocations eliminated.
- Large Trivy parse: ~44% faster, ~74% fewer allocations.
- OS package-name validation: ~97% faster.
- APK/Pacman/RPM validation paths: ~48-52% faster.
- Bulk tag discovery paths: ~68-99% faster depending on strategy.
- Default target-tag resolution: ~99% faster.
- Bulk report index lookup: ~98% faster with zero allocations.

Full benchmark notes and raw progress tracking were kept in `/tmp/perf-copa.md` during development.

## Required CI and security follow-ups

After the branch was updated to current `main`, fresh repository-wide gates exposed failures that were not caused by the hot-path work but blocked this head. The fixes were ported from [PR #1670](https://github.com/project-copacetic/copacetic/pull/1670):

- A fresh vulnerability database reported GO-2026-5856 in the Go 1.25.11 standard library, GO-2026-6061 in gRPC 1.80.0, GO-2026-5158 in OpenTelemetry 1.43.0, and GO-2026-5939 in Claircore 1.5.52. Go, gRPC, and OpenTelemetry were moved only to the first fixed releases (1.25.12, 1.82.1, and 1.44.0 respectively); genproto and other module changes are consequences of those security updates, not performance work.
- Claircore has no fixed release for its finding, and Copa used only `claircore/osrelease`. That dependency was replaced with a bounded, incremental local parser rather than retaining a vulnerable transitive graph.
- The parser's compatibility surface is explicit: it preserves caller cancellation; extracts `NAME`, `ID` (as a fallback), and `VERSION_ID`; handles shell quoting/escapes including escaped apostrophes; and ignores value-syntax errors in fields Copa does not consume. It deliberately rejects malformed consumed fields, invalid UTF-8/control data, and files larger than 1 MiB. Representative Ubuntu, Alpine, Debian, Amazon Linux, CentOS, CBL-Mariner/Azure Linux, RHEL, Rocky, Oracle, and AlmaLinux fixtures plus malformed/edge cases are covered in `pkg/common/os_test.go`.
- Ubuntu 24.04 runners paired Podman 5.x with an obsolete `crun`; the Podman test environment now selects the current runtime for both the initial BuildKit container and inherited Podman/Trivy commands.
- The frontend Dockerfile fallback now matches the Go version in `go.mod`, so local and e2e frontend builds do not use an older compiler.

The gRPC/OpenTelemetry changes were validated with module verification, a zero-reachable-finding `govulncheck` run, and targeted BuildKit/frontend/patch tests; they are also exercised by the exact-head BuildKit/gateway CI matrix.

## Validation

- Exact-head GitHub checks: all 77 PR check contexts passed, including the BuildKit frontend, gateway, Docker/Podman, and multiplatform matrices.
- Focused final-head tests:
  - `go test ./pkg/common ./pkg/report ./pkg/frontend ./pkg/bulk ./pkg/pkgmgr -count=1`
  - targeted `pkg/buildkit` and `pkg/patch` tests
- `go mod verify`
- `govulncheck ./...` (0 reachable vulnerabilities)
- CI-matched `golangci-lint` v2.10.1 (`0 issues`)
- `bash -n` and ShellCheck for the Podman BuildKit environment script
- `git diff --check`
- Targeted benchmark suites for:
  - `pkg/report`
  - `pkg/pkgmgr`
  - `pkg/patch` / `pkg/buildkit`
  - `pkg/bulk`
- `python3 /Users/sozercan/.codex/skills/autoreview/scripts/autoreview --mode local`
  - Final result: `autoreview clean: no accepted/actionable findings reported`

## Notes

- A broad local `make test` reached Docker-backed BuildKit tests but was not authoritative because the local Docker socket became unresponsive. GitHub Actions exercises those paths on Linux runners.

### CODE REVIEW

**Comment by sozercan** [comment]
Addressed both suppressed findings from the Copilot review in cc088882 and 7eb8a668:

- `splitManifestLines` now removes a trailing carriage return from each record, with regression coverage for CRLF, final-CR, blank-line, and embedded-CR inputs.
- report indexing now preserves `encoding/json`-compatible case-insensitive matching for `ArtifactName`, including escaped keys, duplicate ordering, and `null` no-op behavior.

GitHub created no inline review threads for these suppressed findings, so there was nothing available to resolve inline.

File: pkg/common/os.go [hunk 1]
```diff
@@ -19,14 +19,16 @@ type OSInfo struct {
 }
 
 // GetOSInfo extracts OS type and version from os-release data.
-func GetOSInfo(ctx context.Context, osreleaseBytes []byte) (*OSInfo, error) {
-	r := bytes.NewReader(osreleaseBytes)
-	osData, err := osrelease.Parse(ctx, r)
+func GetOSInfo(_ context.Context, osreleaseBytes []byte) (*OSInfo, error) {
+	osData, err := parseOSRelease(osreleaseBytes)
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/common/os.go]: Addressed in 8f2389f5. GetOSInfo now threads the caller context through a bounded incremental parser, checks cancellation before/during/after parsing, avoids bytes.Split allocation, and caps image-controlled input at 1 MiB. Added cancellation, deadline, and oversized-input regression tests.

File: pkg/pkgmgr/rpm.go [hunk 2]
```diff
@@ -1259,15 +1246,23 @@ func validateRPMPackageVersions(updates unversioned.UpdatePackages, cmp VersionC
...
 
-		// Found a match, trim prefix- and drop the .arch suffix to get version string
-		archIndex := strings.LastIndex(lines[lineIndex], "\t")
-		version := strings.TrimPrefix(lines[lineIndex][:archIndex], expectedPrefix)
+		// Found a match, trim prefix and drop the arch suffix to get version string
+		archIndex := strings.LastIndexByte(line, '\t')
+		version := line[nameLen+1:]
+		if archIndex > nameLen {
+			version = line[nameLen+1 : archIndex]
+		}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/pkgmgr/rpm.go]: Addressed in de442a44: RPM result rows now require the architecture delimiter, malformed rows are recorded through the existing error/ignoreErrors path, and a truncated-row regression test was added. The suppressed frontend Go-version finding is addressed in 146ab927 by syncing the Dockerfile fallback to go.mod.

**Review by robert-cronin — APPROVED** [review]
Reviewed at `146ab927`. I ran focused unit tests across the common, report, bulk, and package-manager code, exercised the pure BuildKit/platform helpers, repeated the parser and compatibility boundary tests 10 times, verified dependency and formatting hygiene, and replayed representative benchmarks against the merge base; the results supported the claimed performance improvements without revealing semantic regressions. I did not run Docker, Podman, BuildKit-daemon, or end-to-end tests locally, but all 77 checks on this head are green. Approving the submitted code; the conflicts with `main` still need to be resolved and the resulting head should be rechecked.
